### PR TITLE
EnC: Implements support for adding and removing await/yield return in C# - IDE

### PR DESCRIPTION
**Pull request by tmat**
Since the compiler does not implement VB support yet we need to keep the previous implementation in place and add a new code path for C# in the EnC analyzer. Once we add VB support this code path will be removed.

### CODE REVIEW

File: src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs [hunk 1]
```diff
@@ -377,7 +383,14 @@ private static Label ClassifyStatementSyntax(SyntaxKind kind, SyntaxNode? node,
                     return Label.ForEachStatement;
 
                 case SyntaxKind.UsingStatement:
-                    return Label.UsingStatement;
+                    // We need to distinguish using statements with no or single variable declaration from one with multiple variable declarations.
```
Comment (davidwengier) [inline src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs:386]: Am I reading it wrong, or does the comment not match the code? The code looks like it's distinguishing between no declarations, from single or multiple declarations. Shouldn't it check count of declarations?

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 2]
```diff
@@ -1582,6 +1596,11 @@ private static void AddMatchingActiveNodes(ref List<KeyValuePair<SyntaxNode, Syn
             ImmutableArray<SyntaxNode> oldStateMachineSuspensionPoints,
             ImmutableArray<SyntaxNode> newStateMachineSuspensionPoints)
         {
+            if (SupportsStateMachineUpdates)
```
Comment (davidwengier) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1599]: Nit: This feels like it should be an assert here, and the caller should check it first. Or rename to "TryAdd" maybe?

Though I don't feel too strongly about it.

File: src/Features/VisualBasic/Portable/EditAndContinue/VisualBasicEditAndContinueAnalyzer.vb [hunk 3]
```diff
@@ -2612,6 +2612,16 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.EditAndContinue
...
 #Region "State Machines"
 
+        Protected Overrides ReadOnly Property SupportsStateMachineUpdates As Boolean
+            Get
+                Return False
+            End Get
+        End Property
+
+        Protected Overrides Function IsStateMachineResumableStateSyntax(node As SyntaxNode) As Boolean
+            Return node.IsKind(SyntaxKind.AwaitExpression, SyntaxKind.YieldStatement)
```
Comment (davidwengier) [inline src/Features/VisualBasic/Portable/EditAndContinue/VisualBasicEditAndContinueAnalyzer.vb:2622]: I was surprised to see this doing anything in VB, so I looked for where this method is called, and it isn't. Coming soon?

**Review by davidwengier — APPROVED** [review]

File: src/Features/VisualBasic/Portable/EditAndContinue/VisualBasicEditAndContinueAnalyzer.vb (hunk 3 above)
Comment (tmat) [reply src/Features/VisualBasic/Portable/EditAndContinue/VisualBasicEditAndContinueAnalyzer.vb:2622]: Yes, coming soon.

Comment (tmat) [reply src/Features/VisualBasic/Portable/EditAndContinue/VisualBasicEditAndContinueAnalyzer.vb:2622]: Actually, will be removed for both in coming PR.

File: src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs (hunk 1 above)
Comment (tmat) [reply src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs:386]: "For simplicity we do not match single-declaration to expression, we just treat usings
                    // with declarations entirely separately from usings with expressions."

Comment (tmat) [reply src/Features/CSharp/Portable/EditAndContinue/SyntaxComparer.cs:386]: Meaning - using with expression needs to be different from using with multiple declarations. We might match it with using with a single variable declaration, but it gets complicated and it's not necessary.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 2 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs:1599]: The whole method will be removed by the next PR, which implements VB support.

**Pull request merged**